bitcoin/src/bip32.rs

```diff
@@ -114,37 +114,73 @@ internals::serde_string_impl!(Xpub, "a BIP-32 extended public key");
 /// A child number for a derived key
 #[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
 pub enum ChildNumber {
-    /// Non-hardened key
-    Normal {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
-    /// Hardened key
-    Hardened {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
+    /// A normal child number, indexes 0 to 2147483647 (the first half of all possible children)
+    Normal(u32),
+    /// A hardened child number, indexes 2147483648 to 4294967295 (the second half of all possible children)
+    Hardened(u32),
 }
+
+/// NormalChildNumber is a ChildNumber that is not hardened
+/// made only to be used internally in `derive_pub()` and `NormalChildIterator`
+#[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub struct NormalChildNumber(pub u32);
+
+/// A iterator over the children of a DerivationPath trying to convert them into NormalChildNumber
+/// made only to be used internally in `derive_pub()`
+pub struct NormalDerivationPath {
+    base: Vec<NormalChildNumber>,
+}
+
+impl DerivationPath {
+    /// Try to convert this DerivationPath into a NormalDerivationPath
+    pub fn try_into_normal(&self) -> Result<NormalDerivationPath, Error> {
+        NormalDerivationPath::try_from(self)
+    }
+}
+
+impl TryFrom<&DerivationPath> for NormalDerivationPath {
+    type Error = Error;
+    fn try_from(value: &DerivationPath) -> Result<Self, Self::Error> {
+        let mut base = Vec::with_capacity(value.0.len());
+        for child in value.0.iter() {
+            match child {
+                ChildNumber::Normal(index) => {
+                    base.push(NormalChildNumber(*index));
+                }
```

Comment:
This solves the pop() problem. actually is safe, easier and more readable doing this way.